### ISSUE TITLE
Nonlinear solver regression

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaDiffEq/OrdinaryDiffEq.jl#647]
I think there has been a regression in the nonlinear solver
here's a different example of it
```julia
using ParameterizedFunctions, OrdinaryDiffEq
rober = @ode_def Rober begin
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
end k₁ k₂ k₃
prob = ODEProblem(rober,[1.0;0.0;0.0],(0.0,1e11),(0.04,3e7,1e4))
sol = solve(prob,KenCarp4(),reltol=1e-1,abstol=1e-5)
using Plots
plot(sol,xscale=:log10,tspan=(0.1,1e11))
```

whereas look at https://nbviewer.jupyter.org/github/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/StiffODE/ROBER.ipynb

it used to converge at higher tolerances? Lower tolerances works of course, but I think the convergence check got stricter during v1.0. I think it might be related to qsteady but am not certain.

**Comment by devmotion** [comment]
Maybe you're right but note that in the ROBER notebook the problem is only considered up to `t = 1e5` (and not `t = 1e11` as you wrote above). Still the following plot (from the notebook at high tolerances) suggests that there is a problem with the non-linear solver:
![rober](https://user-images.githubusercontent.com/26102/52310345-61479500-29a3-11e9-8ee4-a9f45dd47d72.png)

**Comment by ChrisRackauckas** [comment]
Bumping this. We definitely shouldn't forget about it.

**Comment by YingboMa** [comment]
```julia
julia> @btime sol = solve($prob,KenCarp4(controller=:_),reltol=1e-4,abstol=1e-8);
  332.661 μs (1595 allocations: 110.17 KiB)

julia> @btime sol = solve($prob,KenCarp4(),reltol=1e-4,abstol=1e-8);
  215.258 μs (1010 allocations: 65.78 KiB)

julia> @btime sol = solve($prob,KenCarp4(controller=:_),reltol=1e-1,abstol=1e-5);
  176.018 μs (883 allocations: 57.56 KiB)

julia> @btime sol = solve($prob,KenCarp4(),reltol=1e-1,abstol=1e-5);
  1.469 ms (3966 allocations: 251.11 KiB)
```
this regression is because of using the predictive controller.

**Comment by ChrisRackauckas** [comment]
That's the performance regression and the diverge? How come the predictive controller is causing a regression? How does qsteady interact with it? I'm not so sure about our defaults there.

**Comment by YingboMa** [comment]
We stoped dispatching to the predictive controller after this commit https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/commit/0dd83c06121077bab7158a6445c33af8d204d12d#diff-77d828dbeb561ecf3e21705f067a485dL3 and it was two years ago. It was fixed in this commit https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/commit/26b2204af707da7c1816ce8fa588574966d5fe3d. So, https://nbviewer.jupyter.org/github/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/StiffODE/ROBER.ipynb only used the standard controller.

**Comment by ChrisRackauckas** [comment]
Yes, I remember that. But that alone fixes the regression? If so, why is the predictive controller performing so poorly?

**Comment by YingboMa** [comment]
I have tried to vary `qsteady_max_default` a bit, but it didn't help. Looks like that we really do need DiffEqStat to investigate this kind of problem.

**Comment by ranocha** [comment]
I'm not completely sure how much this is related, but Hairer and Wanner mention in section 5 of http://www.unige.ch/~hairer/preprints/coimbra.pdf that they use two kinds of step size controllers and take the minimum of both values to get better performance.

**Comment by ChrisRackauckas** [comment]
Well let's remove the regression on release for now and work on the predictive controller.

**Comment by devmotion** [comment]
As an update, the example above converges on master. Moreover, now I get the following plot from the benchmarks mentioned above:
![test_benchmark](https://user-images.githubusercontent.com/26102/53559738-14428300-3b4b-11e9-8876-916a0d060f57.png)
So, the OrdinaryDiffEq algorithms improved but there seems to be a huge regression for radau from ODEInterface :open_mouth:

**Comment by devmotion** [comment]
I don't trust these results, it seems too surprising to me. Can anyone reproduce them? I used the following script which I extracted from https://github.com/JuliaDiffEq/DiffEqBenchmarks.jl/blob/master/StiffODE/ROBER.ipynb:
```julia
using OrdinaryDiffEq, DiffEqProblemLibrary, Sundials, LSODA, ODEInterfaceDiffEq, DiffEqDevTools, Plots

gr()

DiffEqProblemLibrary.ODEProblemLibrary.importodeproblems()
const PROB = DiffEqProblemLibrary.ODEProblemLibrary.prob_ode_rober

# compute test solution
const TEST_SOL = TestSolution(solve(PROB,CVODE_BDF(),abstol=1/10^14,reltol=1/10^14))

function test_high_tol()
    abstols = @. 1/10^(5:8)
    reltols = @. 1/10^(1:4)

    setups = [Dict(:alg=>Rosenbrock23()),
          Dict(:alg=>Kvaerno3()),
          Dict(:alg=>KenCarp4()),
          Dict(:alg=>TRBDF2()),
          Dict(:alg=>KenCarp3()),
    # Dict(:alg=>SDIRK2()), # Removed because it's bad
          Dict(:alg=>radau())]
    names = ["Rosenbrock23" "Kvaerno3" "KenCarp4" "TRBDF2" "KenCarp3" "radau"]
    wp = WorkPrecisionSet(PROB,abstols,reltols,setups;names=names,
                      save_everystep=false,appxsol=TEST_SOL,maxiters=Int(1e5))
    plot(wp)
end
```

**Comment by ranocha** [comment]
I get similar results using your script, master of OrdinaryDiffEq.jl and recent published versions of the other packages:
![figure_1](https://user-images.githubusercontent.com/12693098/53562217-245d6100-3b51-11e9-8afc-7927c3089988.png)

**Comment by ChrisRackauckas** [comment]
That's bizarre, so we should probably look into it more (though I don't know what we'd look at, it may be how the Jacobian is passed to radau that changed?), but our regression looks more than fixed here. In fact, now it looks like we've improved!

**Comment by devmotion** [comment]
It's even more bizarre that this issue affects other packages as well:
```julia
function test_low_tol()
    abstols = @. 1/10^(7:12)
    reltols = @. 1/10^(4:9)

    setups = [Dict(:alg=>Rodas5()),
              Dict(:alg=>Rodas4P()),
              Dict(:alg=>CVODE_BDF()),
              Dict(:alg=>ddebdf()),
              Dict(:alg=>Rodas4()),
              Dict(:alg=>rodas()),
              Dict(:alg=>lsoda()),
              Dict(:alg=>radau())
              ]
    names = ["Rodas5" "Rodas4P" "CVODE_BDF" "ddebdf" "Rodas4" "rodas" "lsoda" "radau"]
    wp = WorkPrecisionSet(PROB,abstols,reltols,setups;names=names,
                      save_everystep=false,appxsol=TEST_SOL,maxiters=Int(1e5))
    plot(wp)
end
```
![benchmark_low_tolerances](https://user-images.githubusercontent.com/26102/53584175-5d162e00-3b83-11e9-92ff-0fd2f1ba8464.png)

**Comment by ChrisRackauckas** [comment]
Yes. Hmm, is `alias_u0` doing something weird on some of the other packages? Are we accidentally changing the problem? Is the Jacobian out of parameterized functions incorrect? Does which package is off change by the ordering of the algorithms?

**Comment by devmotion** [comment]
I had some similar ideas. Checking out the master branch of ODEInterfaceDiffEq and Sundials didn't help at least.

**Comment by devmotion** [comment]
> Does which package is off change by the ordering of the algorithms?

No, I can reproduce the `radau` results by just running a modified version of your example above with `KenCarp4` (which converges now, by the way):
```julia
using ParameterizedFunctions, ODEInterfaceDiffEq
rober = @ode_def Rober begin
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
end k₁ k₂ k₃
prob = ODEProblem(rober,[1.0;0.0;0.0],(0.0,1e11),(0.04,3e7,1e4))
sol = solve(prob,radau(),reltol=1e-1,abstol=1e-5)
using Plots
plot(sol,xscale=:log10,tspan=(0.1,1e11))
```

**Comment by ChrisRackauckas** [comment]
```julia
using ParameterizedFunctions, ODEInterfaceDiffEq, Sundials
alg = CVODE_BDF()
rober = @ode_def Rober begin
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
end k₁ k₂ k₃
rober_bare = @ode_def_bare Rober begin
  dy₁ = -k₁*y₁+k₃*y₂*y₃
  dy₂ =  k₁*y₁-k₂*y₂^2-k₃*y₂*y₃
  dy₃ =  k₂*y₂^2
end k₁ k₂ k₃
prob = ODEProblem(rober,[1.0;0.0;0.0],(0.0,1e11),(0.04,3e7,1e4))
prob_bare = ODEProblem(rober_bare,[1.0;0.0;0.0],(0.0,1e11),(0.04,3e7,1e4))
sol = solve(prob,alg,reltol=1e-1,abstol=1e-5)
sol_bare = solve(prob_bare,alg,reltol=1e-1,abstol=1e-5)
using Plots
plot(sol,xscale=:log10,tspan=(0.1,1e11))
plot(sol_bare,xscale=:log10,tspan=(0.1,1e11))
```

seems to suggest that it's just because it's a difficult problem, and finite differencing is not able to be stable at that tolerance. I think that we should run it on v0.6 and make sure radau diverges with those tolerances there as well, which IIRC it does.

**Comment by devmotion** [comment]
I just checked it on v0.6, and radau diverges with those tolerances.

The reason why those plots look so "bizarre" compared to the plots in the Jupyter notebook and the second comment https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/647#issuecomment-460841222 here is the much longer time span (actually I mentioned that in the second comment but assumed that I had used the longer time span for creating the plot :man_facepalming:). Using the time span `(0, 1e5)` I get:
![test_correct_benchmark](https://user-images.githubusercontent.com/26102/53674112-fc264d00-3c8b-11e9-98b2-6ce8db2f2964.png)

The comparison to the original plot in the notebook and the plot at the beginning of this issue indicates that the regression problem is fixed and there are no apparent problems with radau.
